biome server bugs out on config, CLI works with the same.

I've got Biome setup with PhpStorm. It works perfectly fine from the commandline. 
However the 'integrated server' fails for automated fixes with the following error on startup:

"Biome: The configuration file has errors. Biome will report only parsing errors until the configuration is fixed."

System: W11 with WSL2-ubuntu
PhpStorm: 2024.2.0.1
Biome: 1.8.3

Devdependencies:
        "@biomejs/biome": "1.8.3",
        "@biomejs/cli-linux-x64": "^1.8.3",
        "@biomejs/cli-win32-x64": "^1.8.3",


PhpStorm Biome settings:
+ Automatic Biome configuration
+ Run biome for files: **/*.{js,mjs,mts,cjs,ts,jsx,tsx,cts,json,jsonc,vue,svelte,astro,css}
+ Run format on save.
+ Run safe fixes on save


biome.json in root of the project
```
{
    "$schema": "https://biomejs.dev/schemas/1.8.3/schema.json",
    "formatter": {
        "indentStyle": "space",
        "indentWidth": 4,
        "lineWidth": 120
    },
    "linter": {
        "ignore": ["assets", "core", "layouts", "tests", "types"]
    },
    "vcs": {
        "defaultBranch": "master",
        "clientKind": "git",
        "enabled": true,
        "useIgnoreFile": true
    }
}
```

I don't see anything wrong with this config and running biome format/lint through the CLI works perfectly fine with these settings.

Am I doing something wrong or have I stumbled upon a bug?


With kind regards
Bas.

I will describe my setup and how to reproduce this:

OS: Windows 10/11
WSL2 linux 24.04
Node 22lts (installed in WSL)
pnpm v10 package manager

I am using latest version of biomejs in my project inside wsl2 and Webstorm as IDE. In webstorm settings I have configured that Node.js interpreter is pulled from WSL Ubuntu as well as package manager pnpm. All other plugins like tailwind, typescript, eslint, prettier work fine and read executable from WSL, except biomejs reporting same issue as described here on top.


```
com.intellij.execution.process.ProcessNotCreatedException: Cannot run program "\\wsl.localhost\Ubuntu-24.04\home\ax\code\test\node_modules\@biomejs\biome\bin\biome" (in directory "\\wsl.localhost\Ubuntu-24.04\home\ax\code\test"): CreateProcess error=2, The system cannot find the file specified
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:384)
	at com.intellij.execution.process.OSProcessHandler.startProcess(OSProcessHandler.java:85)
	at com.intellij.execution.process.OSProcessHandler.<init>(OSProcessHandler.java:45)
	at com.intellij.execution.process.CapturingProcessHandler.<init>(CapturingProcessHandler.java:21)
	at com.github.biomejs.intellijbiome.GeneralProcessCommandBuilder$build$4.<init>(GeneralProcessCommandBuilder.kt:55)
	at com.github.biomejs.intellijbiome.GeneralProcessCommandBuilder.build(GeneralProcessCommandBuilder.kt:55)
	at com.github.biomejs.intellijbiome.lsp.BiomeLspServerDescriptor.startServerProcess(BiomeLspServerSupportProvider.kt:70)
	at com.intellij.platform.lsp.impl.connector.Lsp4jServerConnectorStdio.<init>(Lsp4jServerConnectorStdio.kt:16)
	at com.intellij.platform.lsp.impl.LspServerImpl.Q(LspServerImpl.kt:387)
	at com.intellij.platform.lsp.impl.LspServerImpl.Q(LspServerImpl.kt:354)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport$executeOnPooledThread$1.run(AnyThreadWriteThreadingSupport.kt:167)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at com.intellij.util.concurrency.ContextCallable.lambda$call$1(ContextCallable.java:74)
	at com.intellij.util.concurrency.ContextCallable.call(ContextCallable.java:83)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at com.intellij.util.concurrency.ChildContext$runInChildContext$1.invoke(propagation.kt:101)
	at com.intellij.util.concurrency.ChildContext$runInChildContext$1.invoke(propagation.kt:101)
	at com.intellij.util.concurrency.ChildContext.runInChildContext(propagation.kt:107)
	at com.intellij.util.concurrency.ChildContext.runInChildContext(propagation.kt:101)
	at com.intellij.util.concurrency.ContextRunnable.run(ContextRunnable.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:735)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:732)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:732)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.io.IOException: Cannot run program "\\wsl.localhost\Ubuntu-24.04\home\ax\code\test\node_modules\@biomejs\biome\bin\biome" (in directory "\\wsl.localhost\Ubuntu-24.04\home\ax\code\test"): CreateProcess error=2, The system cannot find the file specified

```

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/BiomePackage.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/BiomePackage.kt
@@ -97,6 +97,7 @@ class BiomePackage(private val project: Project) {
                 .getBuilder(binaryPath)
                 .addParameters(listOf(ProcessCommandParameter.Value("--version")))
                 .build()
+                .startProcess()
 
         return runCatching {
             val result = runProcessFuture(processHandler).await()

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/BiomeTargetRun.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/BiomeTargetRun.kt
@@ -3,11 +3,45 @@ package com.github.biomejs.intellijbiome
 import com.github.biomejs.intellijbiome.settings.BiomeSettings
 import com.github.biomejs.intellijbiome.settings.ConfigurationMode
 import com.intellij.execution.ExecutionException
+import com.intellij.execution.configurations.GeneralCommandLine
+import com.intellij.execution.process.CapturingProcessHandler
+import com.intellij.execution.process.OSProcessHandler
+import com.intellij.javascript.nodejs.execution.NodeTargetRun
 import com.intellij.javascript.nodejs.interpreter.NodeJsInterpreterManager
 import com.intellij.javascript.nodejs.interpreter.local.NodeJsLocalInterpreter
 import com.intellij.javascript.nodejs.interpreter.wsl.WslNodeInterpreter
 import com.intellij.lang.javascript.JavaScriptBundle
 import com.intellij.openapi.project.Project
+import com.intellij.util.io.BaseOutputReader
+
+sealed interface BiomeTargetRun {
+    fun startProcess(): OSProcessHandler
+    fun toTargetPath(path: String): String
+    fun toLocalPath(path: String): String
+
+    class Node(private val run: NodeTargetRun) : BiomeTargetRun {
+        override fun startProcess(): OSProcessHandler = run.startProcessEx().processHandler
+        override fun toTargetPath(path: String) = run.convertLocalPathToTargetPath(path)
+        override fun toLocalPath(path: String) = run.convertTargetPathToLocalPath(path)
+    }
+
+    class General(private val command: GeneralCommandLine) : BiomeTargetRun {
+        override fun startProcess() =
+            object : CapturingProcessHandler(command) {
+                override fun readerOptions(): BaseOutputReader.Options {
+                    return object : BaseOutputReader.Options() {
+                        // This option ensures that line separators are not converted to LF
+                        // when the formatter sends e.g., CRLF
+                        override fun splitToLines(): Boolean = false
+                    }
+                }
+            }
+
+        // TODO: custom executable is not supported in WSL yet
+        override fun toTargetPath(path: String) = path
+        override fun toLocalPath(path: String) = path
+    }
+}
 
 class BiomeTargetRunBuilder(val project: Project) {
     fun getBuilder(

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/GeneralProcessCommandBuilder.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/GeneralProcessCommandBuilder.kt
@@ -2,10 +2,7 @@ package com.github.biomejs.intellijbiome
 
 import com.intellij.execution.ExecutionException
 import com.intellij.execution.configurations.GeneralCommandLine
-import com.intellij.execution.process.CapturingProcessHandler
-import com.intellij.execution.process.OSProcessHandler
 import com.intellij.openapi.vfs.VirtualFile
-import com.intellij.util.io.BaseOutputReader
 import java.io.File
 import java.nio.charset.Charset
 import kotlin.io.path.Path
@@ -45,7 +42,7 @@ class GeneralProcessCommandBuilder : ProcessCommandBuilder {
         return this
     }
 
-    override fun build(): OSProcessHandler {
+    override fun build(): BiomeTargetRun {
         val exec = executable ?: throw ExecutionException(BiomeBundle.message("biome.language.server.not.found"))
 
         command.withExePath(exec)
@@ -54,14 +51,6 @@ class GeneralProcessCommandBuilder : ProcessCommandBuilder {
         command.addParameters(parameters)
         charset?.let { command.withCharset(it) }
 
-        return object : CapturingProcessHandler(command) {
-            override fun readerOptions(): BaseOutputReader.Options {
-                return object : BaseOutputReader.Options() {
-                    // This option ensures that line separators are not converted to LF
-                    // when the formatter sends e.g., CRLF
-                    override fun splitToLines(): Boolean = false
-                }
-            }
-        }
+        return BiomeTargetRun.General(command)
     }
 }

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/NodeProcessCommandBuilder.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/NodeProcessCommandBuilder.kt
@@ -1,7 +1,6 @@
 package com.github.biomejs.intellijbiome
 
 import com.intellij.execution.ExecutionException
-import com.intellij.execution.process.OSProcessHandler
 import com.intellij.execution.target.value.TargetValue
 import com.intellij.javascript.nodejs.execution.NodeTargetRun
 import com.intellij.javascript.nodejs.execution.NodeTargetRunOptions.Companion.of
@@ -53,7 +52,7 @@ class NodeProcessCommandBuilder(
         return this
     }
 
-    override fun build(): OSProcessHandler {
+    override fun build(): BiomeTargetRun {
         val exec = executable ?: throw ExecutionException(BiomeBundle.message("biome.language.server.not.found"))
 
         workingDir?.let { builder.setWorkingDirectory(target.path(it)) }
@@ -62,7 +61,6 @@ class NodeProcessCommandBuilder(
         inputFile?.let { builder.setInputFile(TargetValue.fixed(it.path)) }
         charset?.let { builder.charset = it }
 
-        val process = target.startProcessEx()
-        return process.processHandler
+        return BiomeTargetRun.Node(target)
     }
 }

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/ProcessCommandBuilder.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/ProcessCommandBuilder.kt
@@ -1,6 +1,5 @@
 package com.github.biomejs.intellijbiome
 
-import com.intellij.execution.process.OSProcessHandler
 import com.intellij.openapi.vfs.VirtualFile
 import java.nio.charset.Charset
 
@@ -10,5 +9,5 @@ interface ProcessCommandBuilder {
     fun addParameters(params: List<ProcessCommandParameter>): ProcessCommandBuilder
     fun setExecutable(executable: String): ProcessCommandBuilder
     fun setCharset(charset: Charset): ProcessCommandBuilder
-    fun build(): OSProcessHandler
+    fun build(): BiomeTargetRun
 }

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/lsp/BiomeLspServerSupportProvider.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/lsp/BiomeLspServerSupportProvider.kt
@@ -1,9 +1,6 @@
 package com.github.biomejs.intellijbiome.lsp
 
-import com.github.biomejs.intellijbiome.BiomeIcons
-import com.github.biomejs.intellijbiome.BiomePackage
-import com.github.biomejs.intellijbiome.BiomeTargetRunBuilder
-import com.github.biomejs.intellijbiome.ProcessCommandParameter
+import com.github.biomejs.intellijbiome.*
 import com.github.biomejs.intellijbiome.services.BiomeServerService
 import com.github.biomejs.intellijbiome.settings.BiomeConfigurable
 import com.github.biomejs.intellijbiome.settings.BiomeSettings
@@ -11,6 +8,7 @@ import com.intellij.execution.configurations.GeneralCommandLine
 import com.intellij.execution.process.OSProcessHandler
 import com.intellij.openapi.components.service
 import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VfsUtil
 import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.platform.lsp.api.LspServer
 import com.intellij.platform.lsp.api.LspServerSupportProvider
@@ -18,6 +16,7 @@ import com.intellij.platform.lsp.api.ProjectWideLspServerDescriptor
 import com.intellij.platform.lsp.api.customization.LspFormattingSupport
 import com.intellij.platform.lsp.api.lsWidget.LspServerWidgetItem
 import com.intellij.util.SmartList
+import kotlin.io.path.Path
 
 
 @Suppress("UnstableApiUsage") class BiomeLspServerSupportProvider : LspServerSupportProvider {
@@ -44,10 +43,22 @@ import com.intellij.util.SmartList
 }
 
 @Suppress("UnstableApiUsage") private class BiomeLspServerDescriptor(project: Project,
-    val executable: String,
-    val configPath: String?) : ProjectWideLspServerDescriptor(project, "Biome") {
-    private val targetRunBuilder: BiomeTargetRunBuilder
-        get() = BiomeTargetRunBuilder(project)
+    executable: String,
+    configPath: String?) : ProjectWideLspServerDescriptor(project, "Biome") {
+    private val targetRun = run {
+        val params = SmartList<ProcessCommandParameter>(ProcessCommandParameter.Value("lsp-proxy"))
+        if (!configPath.isNullOrEmpty()) {
+            params.add(ProcessCommandParameter.Value("--config-path"))
+            params.add(ProcessCommandParameter.FilePath(configPath))
+        }
+
+        BiomeTargetRunBuilder(project).getBuilder(executable).apply {
+            if (configPath.isNullOrEmpty()) {
+                setWorkingDirectory(configPath)
+            }
+            addParameters(params)
+        }.build()
+    }
 
     override fun isSupportedFile(file: VirtualFile): Boolean {
         return BiomeSettings.getInstance(project).fileSupported(file)
@@ -57,20 +68,14 @@ import com.intellij.util.SmartList
         throw RuntimeException("Not expected to be called because startServerProcess() is overridden")
     }
 
-    override fun startServerProcess(): OSProcessHandler {
-        val params = SmartList<ProcessCommandParameter>(ProcessCommandParameter.Value("lsp-proxy"))
-        if (!configPath.isNullOrEmpty()) {
-            params.add(ProcessCommandParameter.Value("--config-path"))
-            params.add(ProcessCommandParameter.FilePath(configPath))
-        }
+    override fun startServerProcess(): OSProcessHandler =
+        targetRun.startProcess()
 
-        return targetRunBuilder.getBuilder(executable).apply {
-            if (!configPath.isNullOrEmpty()) {
-                setWorkingDirectory(configPath)
-            }
-            addParameters(params)
-        }.build()
-    }
+    override fun getFilePath(file: VirtualFile): String =
+        targetRun.toTargetPath(file.path)
+
+    override fun findLocalFileByPath(path: String): VirtualFile? =
+        VfsUtil.findFile(Path(targetRun.toLocalPath(path)), true)
 
     override val lspGoToDefinitionSupport = false
     override val lspCompletionSupport = null

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/lsp/BiomeLspServerSupportProvider.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/lsp/BiomeLspServerSupportProvider.kt
@@ -8,7 +8,6 @@ import com.intellij.execution.configurations.GeneralCommandLine
 import com.intellij.execution.process.OSProcessHandler
 import com.intellij.openapi.components.service
 import com.intellij.openapi.project.Project
-import com.intellij.openapi.vfs.VfsUtil
 import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.platform.lsp.api.LspServer
 import com.intellij.platform.lsp.api.LspServerSupportProvider
@@ -16,7 +15,6 @@ import com.intellij.platform.lsp.api.ProjectWideLspServerDescriptor
 import com.intellij.platform.lsp.api.customization.LspFormattingSupport
 import com.intellij.platform.lsp.api.lsWidget.LspServerWidgetItem
 import com.intellij.util.SmartList
-import kotlin.io.path.Path
 
 
 @Suppress("UnstableApiUsage") class BiomeLspServerSupportProvider : LspServerSupportProvider {
@@ -75,7 +73,7 @@ import kotlin.io.path.Path
         targetRun.toTargetPath(file.path)
 
     override fun findLocalFileByPath(path: String): VirtualFile? =
-        VfsUtil.findFile(Path(targetRun.toLocalPath(path)), true)
+        super.findLocalFileByPath(targetRun.toLocalPath(path))
 
     override val lspGoToDefinitionSupport = false
     override val lspCompletionSupport = null

